Refactor middleware

kahmali opened the pull request:
Refactor middleware

This is a follow-up of PR #32.

General:
- Ensure all test suites pass when run individually (there were some 
  inconspicuous dependencies between test suites)
- Refactor tests:
  - Namespace all tests into their separate packages (for better
    organization of the output), including a subspace for middleware
  - Rename `expect` function to `waitFor` in all calls to
    testAsyncMulti(), to better reflect its behavior
- Copy .jshintrc into new packages

json-routes:
- Add `JsonRoutes.Middleware` to eventually replace
  `JsonRoutes.middleWare` (since 'middleware' is one word)
- Update change log

rest-accounts-bearer-token:
- Split rest-accounts-bearer-token middleware package into a middleware
  package that parses a standard bearer token (rest-bearer-token-parser)
  and one that validates an auth token and sets the request.userId to
  the authorized Meteor.user's ID (authenticate-user-by-token)
- Attach middleware to RestMiddleware namespace, instead of adding it to
  every route. This allows middleware to be made available by simply
  adding its package. It can later be added to the middleware stack with
  JsonRoutes.Middleware.use(JsonRoutes.Middleware.<middlewareFunction>).
- Move tests to their appropriate package

authenticate-user-by-token:
- Use a strong dependency on `accounts-base` package (was weak)
- Replace our token hashing function with one from the Accounts package

simple-rest:
- Add token and auth middleware in tests to pass auth tests
  - This could be a sign that auth tests may be better suited in the
    corresponding middleware packages
- Update change log

rest-accounts-password:
- Use the latest version of json-routes (1.0.3)
- Add token parsing and auth middleware to the stack (currently added to
  all routes, which is bad - needs fix)
- Move Authentication section of simple:rest docs to this README, since 
  the functionality has been moved here
- Update change log

kahmali commented:
Sorry, forgot to rebase. I need to be awake in 3 hours for my graduation. I'm unfortunately going to have to finish this up right after my graduation. Sorry!

kahmali commented:
OK, rebase is all set. If it's okay to omit the doc updates until an update in the near future, feel free to remove the WIP tag and merge this in. Everything is tested and working as expected (except the load order is still controlled by the order the middleware packages are loaded, and middleware is still being applied over all routes).

aldeed commented on packages/accounts-authorize-user/accounts_authorize_user.js:
```diff
@@ -1,43 +1,32 @@
+  // Works with any middleware that sets req.authToken
+  var token = req.authToken;
   if (! token) {
+    // TODO: Return a 401 Unauthorized error here
```
There should not be an error if authToken isn't set. You may want to add userId parser middleware but not necessarily require a userId (you want to know who's calling if logged in, but allow access anyway if not logged in).

There can be a separate small middleware available that passes an error to `next` if authToken isn't set, in case you want to require an auth token.

kahmali replied:
Makes perfect sense. I'll get rid of this TODO.

aldeed commented on packages/accounts-authorize-user/accounts_authorize_user.js:
```diff
@@ -1,43 +1,32 @@
+  //var tokenExpires = Package["accounts-base"].Accounts._tokenExpiration(token.when);
+  //if (new Date() >= tokenExpires) {
+  //  throw new Meteor.Error("token-expired", "Your login token has expired. Please log in again.");
+  //}
```
You can pass `next` callback into this function, uncomment this, and change the `throw` line to `next({statusCode: 401, error: 'token-expired', reason: 'Your authentication token has expired. Please get a new token.'})`.

kahmali replied:
I'll have to test again, but I think I did something like this and the tests just hung. I think we need to implement error handling middleware before we can propagate errors this way. Don't quote me on any of this. I'll test this out tonight and get back to you on it.

kahmali replied:
As I expected, doing as you suggested caused the tests to hang, because we have no error handling middleware in simple:rest. I tried adding error handling middleware, but I'm unable to force it on the middleware stack after the endpoint (i.e., after any middleware added with `JsonRoutes.add()`). I tried this:
```js
// TODO: Deprecate `middleWare` spelling in favor of `middleware`, since it's one word
JsonRoutes.middleware = JsonRoutes.middleWare = connect();
WebApp.rawConnectHandlers.use(JsonRoutes.middleware);

// List of all defined JSON API endpoints
JsonRoutes.routes = [];

// Save reference to router for later
var connectRouter;

// Register as a middleware
WebApp.rawConnectHandlers.use(connectRoute(function (router) {
  connectRouter = router;
}));

// Error middleware must be added last, to catch errors from prior middleware
JsonRoutes.errorMiddleware = connect();
WebApp.rawConnectHandlers.use(JsonRoutes.errorMiddleware);
```

@stubailo @aldeed Is there anything about that that seems inherently wrong to you? It seemed to follow the existing pattern. 

I separated these changes and I'll push them up on a separate branch soon (unless you think it should be a part of this update).

kahmali replied:
See #43

aldeed commented on packages/accounts-authorize-user/package.js:
```diff
@@ -0,0 +1,29 @@
+Package.describe({
+  name: 'simple:accounts-authorize-user',
```
`accounts-authenticate-user` is probably a more accurate name. Authorization would come later.

kahmali replied:
You're absolutely correct. I'll fix this.

kahmali replied:
Since you can authenticate a user either by token or user and password, would `authenticate-meteor-user-by-token` be a better name? (I figured using `meteor-user` was shorter and more recognizable than `accounts-user`)

aldeed commented on packages/rest-bearer-token-parser/bearer_token_parser.js:
```diff
@@ -0,0 +1,93 @@
+  Fiber(function () {
+    req.authToken = parseHeaders(req) || parseQuery(req) || parseBody(req);
+    next();
+    // TODO: Throw a 401 Unauthorized error
```
We should not do this here. This block of code should be in a separate middleware function that you can add only on paths where you want to require a token. I may want to have all paths try to look for a bearer token, but only require it for some.

kahmali replied:
Also makes perfect sense. Really, really loving how modular this is.

aldeed commented on packages/rest/middleware.js:
```diff
@@ -0,0 +1,15 @@
+ *
+ * TODO: Consider having sub-namespaces for each lifecycle method (e.g., Middleware.pre, Middleware.auth, etc)
+ */
+RestMiddleware = {};
```
It probably makes more sense for RestMiddleware to be defined in json-routes pkg because that package handles all the middleware.

kahmali replied:
Wow. I don't even know how this ended up in this package. We discussed and agreed to put it in `json-routes`, and then apparently that completely slipped my mind and I stuffed it in `simple:rest` where it didn't belong. Sorry about that! I'll move this over to `json-routes` where it belongs.

kahmali replied:
@aldeed @stubailo @raix What do you think about setting up a contract with each middleware package to always declare:

```js
// Load order independent namespacing
RestMiddleware = RestMiddleware || {};

// Add middleware just like before
RestMiddleware.someAwesomeMiddleware = function (req, res, next) {
  // Middleware stuff
};
```
That way every package doesn't need to depend on JsonRoutes just to be integrated with it (so middleware could be used without JsonRoutes without requiring an unnecessary dependency). 

If there's an issue with that I might be overlooking, I guess I'm just torn with having this be a part of JsonRoutes (or any other package for that matter). There's nothing about middleware that makes it reliant on JsonRoutes. It can be used with any Connect router. I don't see why RestMiddleware should be declared in any one package, except maybe its own, as @raix first suggested.

aldeed commented:
@kahmali I added some comments.

Also, once you have switched it to define `RestMiddleware.AccountsAuthenticateUser` as the middleware function, I think the `rest-accounts-password` package should automatically add that middleware by depending on the packages and calling `JsonRoutes.middleWare.use(RestMiddleware.ParseAuthToken)` and `JsonRoutes.middleWare.use(RestMiddleware.AccountsAuthenticateUser)`.

stubailo commented:
@kahmali hey, I noticed you have been changing `function` declarations with `var x = function` function expressions. I'd be against this - I think the prevailing attitude at Meteor is also changing, and most people now prefer the function declaration syntax. One benefit is that you can put all of your helper functions at the bottom of the file, and have the "meat" of the code at the top, rather than having the top of the file filled with tons of helpers.

aldeed commented:
@stubailo, I'm glad to hear Meteor code may be changing this. Functions should have names to help with debugging, so `var foo = function foo()` always seems superfluous to me, especially since they are file-scoped either way in Meteor.

kahmali commented:
_@aldeed: Also, once you have switched it to define `RestMiddleware.AccountsAuthenticateUser` as the middleware function, I think the `rest-accounts-password` package should automatically add that middleware by depending on the packages and calling `JsonRoutes.middleWare.use(RestMiddleware.ParseAuthToken)` and `JsonRoutes.middleWare.use(RestMiddleware.AccountsAuthenticateUser)`._

I can do that, but it might be slightly opinionated to force middleware on the user, right? What if they want the `/login` and `/register` endpoints but don't want to authenticate via standard bearer token? Also, I didn't realize there was no `/logout` endpoint in the rest-accounts-password package - is that deliberate @stubailo, or can I open an issue to add that in?

We need to work on the pattern for applying middleware on specific routes, instead of always applying these on all routes. I'll post the pseudocode we sketched out in a hackpad and create an issue for it so we can start to get the discussion going. 

_@stubailo: hey, I noticed you have been changing function declarations with var x = function function expressions. I'd be against this - I think the prevailing attitude at Meteor is also changing, and most people now prefer the function declaration syntax._

Sorry for putting such a sneaky change in this commit without mentioning it in the commit message or PR. That was evil.  Thanks for catching that @stubailo. What I _thought_ I was doing was removing those functions from the global scope. I incorrectly assumed that because it lacked the `var` keyword the _function declaration_ `function myFunc () {}` was the equivalent of the _function expression_ `myFunc = function () {}`, which would pollute the global scope (one of the bad parts of js). I come from Java originally, where you can always declare helper methods last, so I definitely prefer that style. Luckily, your suggestion led me to a proper understanding of function hoisting in js, so I certainly thank you for that! I'll definitely refactor those. So sorry for my misunderstanding of js there!

aldeed commented:
@kahmali, re "opinionated", yeah, I'm on the fence about that suggestion. But the current release does that (actually the simple:rest pkg does it, so I'm just proposing that the password package do it instead). I was mainly thinking about backwards compatibility (not that anyone is using this yet) and about zero-config. And I think @stubailo's aim was to provide out-of-the-box rest without necessarily needing to understand middleware, etc.

We could actually have yet another package that adds the middleware. It would technically be zero-config because you just have to add the package. But I don't know how people feel about packages with only two lines of code in them. :) It's not the end of the world if the readme has to explain how to add those 2 lines of code.

kahmali commented:
@aldeed I see what you're saying. I just meant something along the lines of providing a config option so that users could override the default auth. Something like:
```js
// Somewhere in the user's meteor app
RestAccountsPassword.configure({
  auth: [
    RestMiddleware.ParseAuthToken, 
    RestMiddleware.AccountsAuthenticateUser
  ]
});

// In the rest-accounts-password-package
if (RestAccountsPassword.authConfigured()) {
  _.each(RestAccountsPassword.config.auth, function (authMiddleware) {
    JsonRoutes.middleWare.use(authMiddleware);
  });
}
else {
  JsonRoutes.middleWare.use(RestMiddleware.ParseAuthToken);
  JsonRoutes.middleWare.use(RestMiddleware.AccountsAuthenticateUser);
}
```

I'm not at all sold on that pattern though, so I'll just go ahead and "hard-code" them for now. This may need to be a part of a larger discussion, because I'm not sure where @stubailo stands on configuration in simple:rest. I love the idea of zero-config, out-of-the-box REST functionality, but at some point, a user (myself included) is going to need fine-grained control over their API, and so I would hope that one of our overarching goals could be "sensible defaults with total customization." 

The one major difference between the original implementation and this new one is that we're now (inappropriately) adding that middleware onto every route (not just the api routes, but every route on the server). I'll create the hackpad and issue to get the discussion going on api namespacing. This should fall within the scope. The entire api could be namespaced (there could be a layer of "subspaces" under that) and then we could apply middleware only on api routes.

kahmali commented on packages/accounts-authorize-user/accounts_authorize_user.js:
```diff
@@ -1,43 +1,32 @@
-    );
-  }
+  // TODO: Check token expiration?
+  //var tokenExpires = Package["accounts-base"].Accounts._tokenExpiration(token.when);
```
@stubailo @aldeed In the hashToken function on line 45, a check is made to verify that the `token` is a `String` (and the value of `token` is also being set in middleware as a `String` via `req.authToken`), so why are we relying on a value at token.when?

kahmali commented:
Just pushed an update that should address all the earlier comments. I'm going to remove the WIP label, since the documentation is all there and I believe this is a "complete" update. Of course, it may still need a little work, since I wasn't able to completely resolve the error issue. I will push the error handling middleware onto a separate branch, so we can start the discussion around that.

kahmali commented:
@aldeed True indeed. Fixed locally. Will push with any other recommended changes.

aldeed commented:
@kahmali, a couple more comments added on the commit, but looks pretty good.

Thinking about naming, would it make more sense for `JsonRoutes.middleware` to be the place where middleware functions are attached (instead of having `RestMiddleware` object)? Based on the way connect uses naming, it seems like the API should be

```js
JsonRoutes.connect.use(JsonRoutes.middleware.ParseAuthToken);
```

Or if we provide multiple connection points, something like:

```js
JsonRoutes.connect.preAuth.use(JsonRoutes.middleware.ParseAuthToken);
JsonRoutes.connect.auth.use(JsonRoutes.middleware. AccountsAuthenticateUser);
JsonRoutes.connect.postAuth.use(MyXMLtoJSONBodyConversionMiddleware);
```

kahmali commented:
_@aldeed: Thinking about naming, would it make more sense for `JsonRoutes.middleware` to be the place where middleware functions are attached (instead of having `RestMiddleware` object)? Based on the way connect uses naming, it seems like the API should be_

```js
JsonRoutes.connect.use(JsonRoutes.middleware.ParseAuthToken);
```

_Or if we provide multiple connection points, something like:_

```js
JsonRoutes.connect.preAuth.use(JsonRoutes.middleware.ParseAuthToken);
JsonRoutes.connect.auth.use(JsonRoutes.middleware. AccountsAuthenticateUser);
JsonRoutes.connect.postAuth.use(MyXMLtoJSONBodyConversionMiddleware);
```

I like this a lot. If @stubailo agrees I think we should go this route. While we're on the topic of naming in `JsonRoutes`, I think we should reconsider the entire package name. It won't be long until it supports a lot more than just JSON responses. Should we get ahead of the curve and change this name to something more generic, like `SimpleRoutes` or `SimpleRestRoutes`, or something along those lines?

stubailo commented:
I totally agree with JsonRoutes.middleware. Also, have we actually gone beyond JsonRoutes yet? AFAIK everything we are doing is returning JSON for now.

aldeed commented:
I think keeping JsonRoutes name is fine for now.

kahmali commented:
@stubailo @aldeed Please review. 

Just pushed the latest. It should address all the feedback. This PR should be nearing completion.

stubailo commented on packages/authenticate-user-by-token/README.md:
```diff
@@ -0,0 +1,36 @@
+#simple:authenticate-user-by-token
```
Put a space after the hash

stubailo commented on packages/authenticate-user-by-token/README.md:
```diff
@@ -0,0 +1,36 @@
+#simple:authenticate-user-by-token
+
+SimpleRest middleware for validating a Meteor.user's login token
```
I would say these are more like "connect middleware" or "JsonRoutes middleware" or maybe just middleware? There's nothing about them that would require using simple:rest.

aldeed replied:
"Connect middleware" is probably most accurate.

kahmali replied:
Oops. I should have converted the name over when I moved the middleware namespace to `JsonRoutes.Middleware`. I think it should be "JsonRoutes middleware." My hope is that "JsonRoutes middleware" would imply "Connect middleware" the same way "Express middleware" did for a long time (before they stopped relying on Connect). So the idea is that _all_ JsonRoutes Middleware is Connect middleware, with the additional interface requirements met (pretty much just the fact that the middleware is added to the `JsonRoutes.Middleware` namespace). That way we can just specify from the `json-routes` docs that all middleware also works with any framework that supports Connect middleware.

stubailo commented on packages/rest-bearer-token-parser/package.js:
```diff
@@ -1,8 +1,8 @@
 Package.describe({
-  name: 'simple:rest-accounts-bearer-token',
+  name: 'simple:rest-bearer-token-parser',
```
I think we should fiddle with these names a little, but we can definitely do that after merging, right before we release. For example, this name might be "simple:bearer-token-parser-middleware" since it's not really related to REST.

kahmali replied:
I agree. I've started a hackpad for our Middleware Patterns that I'll share soon. We can discuss the naming conventions for middleware packages and functions there. The intention, in the back of my mind, is separating layers that would be REST or DDP dependent, and then having a layer of shared middleware. This is definitely a part of a larger discussion. 

I _was_ under the impression that this was REST-specific, since it's parsing the bearer token from the headers. I (shamefully) don't understand the DDP spec enough to know if it uses headers. Nothing I've come across has made it appear that HTTP headers were a part of the DDP spec, but I could be completely off base here. Like you said, I think these names are acceptable for this PR, but we should definitely resolve them before release.

aldeed replied:
I think @stubailo means it is specific to HTTP requests but not specific to rest endpoints. Could be used by other non-restful server-side HTTP routing.

stubailo commented:
@kahmali I think this looks awesome. We should fiddle a little bit with the readmes, summaries, names, and version numbers before releasing, but I'm happy with merging this.

aldeed commented:
@kahmali @stubailo, the bearer token parsing middleware ended up looking very similar to this: https://www.npmjs.com/package/express-bearer-token

Should we just use/recommend that middleware or do a lightweight wrapper package around it to export for Meteor?

stubailo commented:
@aldeed still not really happy about parsing it from the body... but maybe it's not such a big deal. What are the chances someone will want to insert a document with a field called access_token?

kahmali commented:
@stubailo @aldeed I've addressed those final minor suggested changes and rebased. Dare I say this PR is ready to be merged?

It looks like Express decided to stick with the RFC standard. I'm not against providing a wrapper around that package. I think devs should just need to understand the consequences of passing metadata like that in the request body. Somebody could always add a layer of middleware for removing the field, as I think one of you may have already suggested. And then people get a true "standard bearer token" parser.

stubailo commented:
@kahmali ok going to merge manually now

stubailo commented:
merged: 82250cb8109c86b680618ff8aeb390158302df2d

stubailo commented:
@kahmali serves me right for not running the code before merging, but it seems you forgot to add the file `middleware.js`.

aldeed commented:
@stubailo, regarding parsing from body, there is also the fact that it's supposed to be urlencoded post rather than json. I'm not even sure exactly how we could support both. It's probably better if we don't support post body, or at least don't document that we do.

kahmali commented:
@aldeed I would lean towards just outright not supporting it. I think it could lead to some confusion in the unlikely event that you're using this middleware and have an `access_token` field in your request body without any valid bearer token header or query param. Also, would we really have to do anything special to handle both urlencoded and json POSTs? Isn't that already a feature of the `Connect.bodyParser` middleware we're using?

aldeed commented:
@kahmali I meant I don't think it's possible for the same request to be both urlencoded and json, as far as sending auth token along with your document to insert or whatever. Agree we should remove support for auth in body.
